src/pmfs/vos_target_fs.c

```diff
@@ -0,0 +1,355 @@
+/**
+ * (C) Copyright 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+ #include <stdio.h>
+ #include <string.h>
+ #include <errno.h>
+ #include <sys/types.h>
+ #include <sys/stat.h>
+ #include <sys/mman.h>
+ #include <stdlib.h>
+ #include <unistd.h>
+ #include <daos_srv/vos.h>
+ #include <daos_fs_sys.h>
+ #include <spdk/env.h>
+ #include "vos_target_fs.h"
+ #include "vos_tasks.h"
+ #include "vos_target_engine.h"
+ #include <pmfs/pmfs.h>
+
+char dfs_pmem_file[PATH_MAX];
+static pthread_mutex_t vos_fs_cmd_lock = PTHREAD_MUTEX_INITIALIZER;
+D_LIST_HEAD(g_task_ring);
+
+static inline double
+dts_time_now(void)
+{
+	struct timeval	tv;
+
+	gettimeofday(&tv, NULL);
+	return (tv.tv_sec + tv.tv_usec / 1000000.0);
+}
+
+const char *cmd_string[] = {"none", "update", "fetch", "punch", "get_num_dkeys", "list_dkeys"};
+
+static int
+vos_obj_get_info_dkeys(daos_handle_t coh, daos_unit_oid_t oid,
+		       enum task_op opc, uint32_t *nr, uint64_t *len,
+		       daos_key_desc_t *kds, void *buf)
+{
+	struct scan_context ctx = { 0 };
+	struct pmfs_obj_info *uoi = NULL;
+	int rc = 0;
+
+	ctx.cur_cont.tsc_coh = coh;
+	ctx.uoi.oid = oid;
+	ctx.uoi.nr = *nr;
+	ctx.uoi.len = *len;
+
+	D_ALLOC(uoi, sizeof(struct pmfs_obj_info));
+	D_ASSERT(uoi != NULL);
+
+	rc = pmfs_scan_cont(&ctx, &uoi, opc);
+	if (rc != 0) {
+		D_ERROR("Object get dkeys info failed: " DF_RC "\n",
+			DP_RC(rc));
+		return rc;
+	}
+
+	*nr = uoi->nr;
+	*len = uoi->len;
+
+	return 0;
+}
+
+static int
+vos_obj_list_info_dkeys(daos_handle_t coh, daos_unit_oid_t oid,
+			enum task_op opc, uint32_t *nr, uint64_t *len,
+			daos_key_desc_t *kds, void *buf)
+{
+	struct scan_context ctx = { 0 };
+	struct pmfs_obj_info *uoi = NULL;
+	int rc = 0;
+
+	ctx.cur_cont.tsc_coh = coh;
+	ctx.uoi.oid = oid;
+	ctx.uoi.nr =  *nr;
+	ctx.uoi.len = *len;
+	ctx.uoi.kds = kds;
+	ctx.uoi.buf = buf;
+
+	D_ALLOC(uoi, sizeof(struct pmfs_obj_info));
+	D_ASSERT(uoi != NULL);
+
+	rc = pmfs_scan_cont(&ctx, &uoi, opc);
+	if (rc != 0) {
+		D_ERROR("Object get dkeys info failed: " DF_RC "\n",
+			DP_RC(rc));
+		return rc;
+	}
+
+	*nr = uoi->nr;
+	*len = uoi->len;
+	kds = uoi->kds;
+	buf = uoi->buf;
+
+	return 0;
+}
+
+
+struct vos_fs_ult_arg {
+	void *args;
+	enum task_op opc;
+	double *duration;
+	int status;
+};
+
+static int
+_vos_parse_commands(enum task_op opc, void *union_args)
+{
+	int rc = 0;
+
+	if (opc == VOS_OBJ_UPDATE) {
+		struct vos_client_obj_rw_args *args = union_args;
+
+		rc = vos_obj_update(args->coh, args->oid,
```

Comment:
we are potentially leaving some performance on the table.   vos_obj_update may yield the ULT internally (argobots) when doing the copy to NVMe but if we don't use argobots, that essentially means these calls are syncrhonous.  Is this expected?

Reply:
Thanks @jolivier23 , my initial draft codes have two path, another one is using (argobots) ULT. But that met a hang issue.
For the aim to verify the threading model is workable, so using regular thread instead.  I will try to fix that and update.